bundles/pubsub/pubsub_admin_udp/src/pubsub_udp_topic_sender.c

```diff
@@ -0,0 +1,516 @@
+    bool is_broadcast = pubsub_utils_url_is_broadcast(urlInfo->hostname);
+    pubsub_utils_url_free(urlInfo);
+    if (!is_multicast && !is_broadcast) {
+        L_DEBUG("[PSA udp] TopicSender %s/%s disconnect from udp url %s",
+                sender->scope == NULL ? "(null)" : sender->scope,
+                sender->topic,
+                url);
+        pubsub_sktHandler_disconnect(sender->socketHandler, connectUrl);
+    }
+}
+
+static void *psa_udp_getPublisherService(void *handle, const celix_bundle_t *requestingBundle,
+                                         const celix_properties_t *svcProperties __attribute__((unused))) {
+    pubsub_udp_topic_sender_t *sender = handle;
+    long bndId = celix_bundle_getId(requestingBundle);
+
+    celixThreadMutex_lock(&sender->boundedServices.mutex);
+    psa_udp_bounded_service_entry_t *entry = hashMap_get(sender->boundedServices.map, (void *) bndId);
+    if (entry != NULL) {
+        entry->getCount += 1;
+    } else {
+        entry = calloc(1, sizeof(*entry));
+        entry->getCount = 1;
+        entry->parent = sender;
+        entry->bndId = bndId;
+        entry->service.handle = entry;
+        entry->service.localMsgTypeIdForMsgType = psa_udp_localMsgTypeIdForMsgType;
+        entry->service.send = psa_udp_topicPublicationSend;
+        hashMap_put(sender->boundedServices.map, (void *) bndId, entry);
+    }
+    celixThreadMutex_unlock(&sender->boundedServices.mutex);
+
+    return &entry->service;
+}
+
+static void psa_udp_ungetPublisherService(void *handle, const celix_bundle_t *requestingBundle,
+                                          const celix_properties_t *svcProperties __attribute__((unused))) {
+    pubsub_udp_topic_sender_t *sender = handle;
+    long bndId = celix_bundle_getId(requestingBundle);
+
+    celixThreadMutex_lock(&sender->boundedServices.mutex);
+    psa_udp_bounded_service_entry_t *entry = hashMap_get(sender->boundedServices.map, (void *) bndId);
+    if (entry != NULL) {
+        entry->getCount -= 1;
+    }
+    if (entry != NULL && entry->getCount == 0) {
+        //free entry
+        hashMap_remove(sender->boundedServices.map, (void *) bndId);
+        free(entry);
+    }
+    celixThreadMutex_unlock(&sender->boundedServices.mutex);
+
+}
+
+static int
+psa_udp_topicPublicationSend(void *handle, unsigned int msgTypeId, const void *inMsg, celix_properties_t *metadata) {
+    psa_udp_bounded_service_entry_t *bound = handle;
+    pubsub_udp_topic_sender_t *sender = bound->parent;
+    const char* msgFqn;
+    int majorVersion;
+    int minorVersion;
+    celix_status_t status = pubsub_serializerHandler_getMsgInfo(sender->serializerHandler, msgTypeId, &msgFqn, &majorVersion, &minorVersion);
+
+    if (status != CELIX_SUCCESS) {
+        L_WARN("Cannot find serializer for msg id %u for serializer %s", msgTypeId,
+               pubsub_serializerHandler_getSerializationType(sender->serializerHandler));
```

Comment:
Memory leak in psa_tcp is reported by CI and fixed by: https://github.com/apache/celix/pull/403